File: src/main/java/xyz/mizarc/persistentitems/commands/persistentitems/HideCommand.java
package xyz.mizarc.persistentitems.commands.persistentitems;

import co.aikar.commands.BaseCommand;
import co.aikar.commands.annotation.*;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;
import xyz.mizarc.persistentitems.DatabaseConnection;
import xyz.mizarc.persistentitems.Item;
import xyz.mizarc.persistentitems.ItemContainer;
import xyz.mizarc.persistentitems.PersistentItems;

import java.util.Arrays;

@CommandAlias("persistentitems|pitems|pi")
public class HideCommand extends BaseCommand {

    @Dependency
    private PersistentItems plugin;

    @Dependency
    ItemContainer itemContainer;

    @Subcommand("hide")
    @CommandPermission("persistentitems.command.hide")
    @CommandCompletion("@pitems @player")
    @Syntax("<item> [player]")
    public void onHide(CommandSender sender, String itemId, @Optional Player specifiedPlayer) {
        // Error if console is trying to use this without specifying a player
        if (!(sender instanceof Player) && specifiedPlayer == null) {
            sender.sendMessage("You must specify the player argument as the console");
            return;
        }

        // Error if persistent item is not active
        if (itemContainer.getItem(itemId) == null) {
            sender.sendMessage("Item " + sender + " does not exist");
            return;
        }

        // Add item to specified player's inventory unless player already has it
        if (specifiedPlayer != null) {
            if (!removeFromInventory(specifiedPlayer.getInventory(), itemId)) {
                sender.sendMessage("Item " + itemId + " is not in " + specifiedPlayer + "'s inventory");
                return;
            }
            addToDatabase(specifiedPlayer.getUniqueId().toString(), itemId);
            sender.sendMessage("Item " + itemId + " has been removed from " + specifiedPlayer + "'s inventory");
            return;
        }

        // Remove item from own inventory unless you don't have it
        Player player = (Player) sender;
        if (!removeFromInventory(player.getInventory(), itemId)) {
            sender.sendMessage("Item " + itemId + " is not in your inventory");
            return;
        }
        addToDatabase(player.getUniqueId().toString(), itemId);
        sender.sendMessage("Item " + itemId + " has been removed from your inventory");
    }

    private boolean removeFromInventory(PlayerInventory inventory, String itemName) {
        ItemContainer container = plugin.getItemContainer();
        Item item = container.getItem(itemName);
        ItemStack itemStack = item.getItemStack(plugin);

        // False if the inventory doesn't have the item
        if (inventory.contains(itemStack)) {
            return false;
        } else if (inventory.getItemInOffHand().equals(itemStack)) {
            return false;
        } else if (Arrays.asList(inventory.getArmorContents()).contains(itemStack)) {
            return false;
        }

        // True if the inventory has the item and remove it
        inventory.remove(itemStack);
        return true;
    }

    private void addToDatabase(String playerId, String itemName) {
        DatabaseConnection database = new DatabaseConnection(plugin);
        database.addHidden(playerId, itemName, "global");
        database.closeConnection();
    }
}

File: src/main/kotlin/xyz/mizarc/persistentitems/commands/persistentitems/HideCommand.kt
package xyz.mizarc.persistentitems.commands.persistentitems

import co.aikar.commands.ConditionFailedException
import co.aikar.commands.MessageKeys
import co.aikar.commands.annotation.*
import co.aikar.commands.annotation.Optional
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.inventory.PlayerInventory
import xyz.mizarc.persistentitems.Item
import xyz.mizarc.persistentitems.commands.PersistentItemsCommand
import java.util.*

@CommandAlias("persistentitems|pitems|pi")
class HideCommand : PersistentItemsCommand() {

    @Subcommand("hide")
    @CommandPermission("persistentitems.command.hide")
    @CommandCompletion("@pitems @players")
    @Syntax("<item> [player]")
    fun onHide(sender: CommandSender, name: String, @Optional player: Player) {
        // Error if persistent item is not active
        val item = itemRepo.getByName(name).firstOrNull()
        if (item == null) {
            sender.sendMessage("Item $sender does not exist")
            return
        }

        // Forward command to 'others' version if player argument is specified
        if (player != null) {
            onHideOthers(sender, item, player)
            return
        }

        // Error if console is trying to use this without specifying a player
        if (sender !is Player) {
            sender.sendMessage("You must specify the player argument as the console")
            return
        }

        // Remove item from own inventory unless you don't have it
        val selfPlayer = sender
        if (!removeFromInventory(selfPlayer.inventory, item)) {
            sender.sendMessage("Item ${item.id} is not in your inventory")
            return
        }
        playerItemsRepo.add(selfPlayer, item)
        sender.sendMessage("Item ${item.id} has been removed from your inventory")
    }

    private fun onHideOthers(sender: CommandSender, item: Item, player: Player) {
        // Error if player doesn't have the 'others' permission
        if (!sender.hasPermission("persistentitems.command.hide.others")) {
            throw ConditionFailedException(MessageKeys.PERMISSION_DENIED_PARAMETER)
        }

        // Remove item from specified player's inventory unless player doesn't have it
        if (!removeFromInventory(player.inventory, item)) {
            sender.sendMessage("Item ${item.id} is not in ${player.displayName()}'s inventory")
            return
        }

        playerItemsRepo.add(player, item)
        sender.sendMessage("Item ${item.id} has been removed from ${player.displayName()}'s inventory")
    }

    private fun removeFromInventory(inventory: PlayerInventory, item: Item): Boolean {
        val itemSlots: List<Int> = persistencyService.getSlotsWithItems(inventory, item.id) ?: return false
        if (itemSlots.isEmpty()) {
            return false
        }

        // Remove instances of persistent items from inventory
        for (slot in itemSlots) {
            inventory.setItem(slot, null)
        }
        return true
    }
}
